File: backend-express/src/models/CallLog.js
const mongoose = require('mongoose');

const callLogSchema = new mongoose.Schema(
  {
    caller: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    receiver: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    type: {
      type: String,
      enum: ['voice', 'video'],
      required: true,
    },
    status: {
      type: String,
      enum: ['missed', 'answered', 'declined', 'ended'],
      default: 'missed',
    },
    duration: { type: Number, default: 0 }, // seconds
    startedAt: { type: Date, default: null },
    endedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

callLogSchema.index({ caller: 1, createdAt: -1 });
callLogSchema.index({ receiver: 1, createdAt: -1 });

module.exports = mongoose.model('CallLog', callLogSchema);

File: app/lib/core/models/call_log.dart
class CallLog {
  final String id;
  final CallUser caller;
  final CallUser receiver;
  final String type; // 'voice' | 'video'
  final String status; // 'missed' | 'answered' | 'declined' | 'ended'
  final int duration; // seconds
  final bool isOutgoing;
  final CallUser otherUser;
  final DateTime createdAt;

  const CallLog({
    required this.id,
    required this.caller,
    required this.receiver,
    required this.type,
    required this.status,
    required this.duration,
    required this.isOutgoing,
    required this.otherUser,
    required this.createdAt,
  });

  bool get isMissed => status == 'missed' || status == 'declined';
  bool get isVideo => type == 'video';

  String get durationLabel {
    if (duration == 0) return '';
    final m = duration ~/ 60;
    final s = duration % 60;
    return '${m.toString().padLeft(2, '0')}:${s.toString().padLeft(2, '0')}';
  }

  factory CallLog.fromJson(Map<String, dynamic> json) {
    final callerData = json['caller'] as Map<String, dynamic>? ?? {};
    final receiverData = json['receiver'] as Map<String, dynamic>? ?? {};
    final otherData = json['otherUser'] as Map<String, dynamic>? ?? callerData;

    return CallLog(
      id: json['_id'] ?? json['id'] ?? '',
      caller: CallUser.fromJson(callerData),
      receiver: CallUser.fromJson(receiverData),
      type: json['type'] ?? 'voice',
      status: json['status'] ?? 'missed',
      duration: json['duration'] ?? 0,
      isOutgoing: json['isOutgoing'] ?? false,
      otherUser: CallUser.fromJson(otherData),
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt']) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class CallUser {
  final String id;
  final String nickname;
  final String? avatarUrl;

  const CallUser({
    required this.id,
    required this.nickname,
    this.avatarUrl,
  });

  factory CallUser.fromJson(Map<String, dynamic> json) => CallUser(
        id: json['_id'] ?? json['id'] ?? '',
        nickname: json['nickname'] ?? '',
        avatarUrl: json['avatarUrl'],
      );
}
